janitor/functions/min_max_scale.py

```diff
@@ -23,23 +23,18 @@ def min_max_scale(
     df: pd.DataFrame,
     feature_range: tuple[int | float, int | float] = (0, 1),
     column_name: str | int | list[str | int] | pd.Index = None,
+    entire_data: bool = False,

```

Comment:
I don't think `entire_data` is a better name.

Reply:
I agree, though I also need a bit of time and space to think of a better name. I wonder if others on the dev team have ideas? @pyjanitor-devs/core-devs

Reply:
not sure of a good name to use. maybe `keep` like in Pandas?

Reply:
You mean 'keep' its value could be 'column' or 'all'?

But this variable is better as a boolean type.

Reply:
yea, horrible parameter name. maybe `scale_all` or `scale_all_columns`?

Reply:
Hmmm, @Zeroto521, on second thought, I think we need a bit better definition of the semantics. `min_max_scale` currently operates with the assumption of operating on one column, so the use of `column_name` makes sense here. Below is my attempt at reasoning through the multiple ways we could use `min_max_scale`.

- Scale one column.
- Scale multiple columns independently.
- Scale multiple columns, but jointly (so they are all scaled to the same min and max)

`entire_data` is a special case of scaling multiple columns _jointly_. 

Since we're working on this function, it's a good chance to change the API to be flexible yet also sensible. What if the API was, instead the following?

```python
def min_max_scale(df, feature_range, column_names: Iterable[Hashable] | Callable, jointly: bool):
```

If `jointly` is True, then the `column_names` provided are jointly scaled; otherwise, they are not. 

I wanted to point out a new behaviour that we might be able to support across the rest of the API -- by making `column_names` accept a Callable that has the signature:

```python
def column_names_callable(df) -> Iterable[Hashable]:
```

we can enable `min_max_scale` on all columns by doing:

```python
df.min_max_scale(feature_range=(0, 1), column_names = lambda df: df.columns, jointly=True)
```

This is pretty concise without resorting to needing to maintain string mappings for special-case behaviour.

I'm glad we didn't rush to merge this PR, giving us the time and space to think clearly about the semantics of the API. 

@samukweku and @Zeroto521 what do you think about this?

Reply:
> I wanted to point out a new behaviour that we might be able to support across the rest of the API -- by making column_names accept a Callable that has the signature:
> ```python
> def column_names_callable(df) -> Iterable[Hashable]:
> ```

I'm sorry I still can't get the point why `column_names` could accept a callback type argument.

> we can enable min_max_scale on all columns by doing:
> ```python
> df.min_max_scale(feature_range=(0, 1), column_names = lambda df: df.columns, jointly=True)
> ```

To scale all columns, I thought we could use the default argument as `None` for `column_names` without no more inputting.

```python
df.min_max_scale(feature_range=(0, 1), column_names=None, jointly=True)
```

Are there more examples to show the importance of the callback type?

Reply:
Thanks for the comments, @Zeroto521! 

Regarding why we might want to allow `column_names` to be a Callable, I had the idea that it helps support being explicit over implicit, which is in the Zen of Python. Setting `column_names=None` makes selecting all columns implicit, whereas setting `column_names=lambda df: df.columns` makes selecting all columns explicit. In addition, it allows the selection of arbitrary subsets of column names programmatically, without needing to hard-code those names.

On further thought, I can see how `column_names=None` actually follows the pattern established in other places in the library, so I think, for now, we can:

1. Use `column_names=None` to imply selection of all columns, and
2. Talk more about `column_names: Callable` in the issue tracker, deferring the implementation till later.

What do you all think about `jointly=True` as the keyword for triggering whether to _independently_ scale each column or to _jointly_ scale all columns specified in `column_names`? @Zeroto521 if you're in agreement with the keyword argument, then I think, let's get that specified in this PR, then we can close out the PR!

Reply:
I totally agree with using `jointly`.

About `column_names` whether could receive a callable type or not.
I can understand now. `column_names=lambda df: df.columns` is an implicit style and also a trick.

| Select columns                 | Using callable type                                                   | Using Iterable type | `df.columns`                 |
| ------------------------------ | --------------------------------------------------------------------- | ------------------- | ---------------------------- |
| Select the first three columns | `column_names=lambda df: df.columns[:3]`                              | `['a', 'b', 'c']`   | pd.Index(list('abcde'))      |
| Select str type columns        | `column_names=lambda df: [i for i in df.columns if instance(i, str)]` | `['a', 'b', 'c']`   | pd.Index(['a', 'b', 'c', 1]) |

As you said, we can put it aside at present.
Once the parameter `column_names` of `min_max_scale` could receive, other functions also need to do the same thing.

Reply:
More discussions move to #1115